src/execution/operator/projection/physical_tableinout_function.cpp

```diff
@@ -52,15 +56,152 @@ unique_ptr<GlobalOperatorState> PhysicalTableInOutFunction::GetGlobalOperatorSta
 	}
+}
+
+void PhysicalTableInOutFunction::AddProjectedColumnsFromConstantMapping(idx_t map_idx, DataChunk &input,
+                                                                        DataChunk &intermediate, DataChunk &out) const {
+	auto &mapping_column = intermediate.data[map_idx];
+	D_ASSERT(mapping_column.GetVectorType() == VectorType::CONSTANT_VECTOR);
+
+	UnifiedVectorFormat mapping_vector_data;
+	mapping_column.ToUnifiedFormat(intermediate.size(), mapping_vector_data);
+	auto mapping_data = (sel_t *)mapping_vector_data.data;
+
+	// Add the projected columns, and apply the selection vector
+	for (idx_t project_idx = 0; project_idx < projected_input.size(); project_idx++) {
+		auto source_idx = projected_input[project_idx];
+		D_ASSERT(source_idx < input.data.size());
+		auto target_idx = map_idx + project_idx;
+
+		auto &target_column = out.data[target_idx];
+		auto &source_column = input.data[source_idx];
+
+		ConstantVector::Reference(target_column, source_column, mapping_data[0], input.size());
+	}
+}
+
+void PhysicalTableInOutFunction::AddProjectedColumnsFromFlatMapping(idx_t map_idx, DataChunk &input,
+                                                                    DataChunk &intermediate, DataChunk &out) const {
+	auto &mapping_column = intermediate.data[map_idx];
+	D_ASSERT(mapping_column.GetVectorType() == VectorType::FLAT_VECTOR);
+
+	UnifiedVectorFormat mapping_data;
+	mapping_column.ToUnifiedFormat(intermediate.size(), mapping_data);
+	D_ASSERT(mapping_data.validity.AllValid());
+	auto mapping_array = (sel_t *)mapping_data.data;
+
+	// We can directly use this column as a selection vector
+	SelectionVector sel_vec(mapping_array);
+
+	// Add the projected columns, and apply the selection vector
+	for (idx_t project_idx = 0; project_idx < projected_input.size(); project_idx++) {
+		auto source_idx = projected_input[project_idx];
+		D_ASSERT(source_idx < input.data.size());
+		auto target_idx = map_idx + project_idx;
+
+		auto &target_column = out.data[target_idx];
+		auto &source_column = input.data[source_idx];
+
+		target_column.Slice(source_column, sel_vec, intermediate.size());
+		// Since our selection vector is using temporary allocated data, we need to
+		// immediately flatten this column, so we don't run the risk of the dictionary vector
+		// outliving the selection vector
+		target_column.Flatten(intermediate.size());
+	}
+}
+
+OperatorResultType PhysicalTableInOutFunction::ExecuteWithMapping(ExecutionContext &context, DataChunk &input,
+                                                                  DataChunk &chunk, TableInOutLocalState &state,
+                                                                  TableFunctionInput &data) const {
+	// Create a duplicate of 'chunk' that contains one extra column
+	// this column is used to register the relation between input tuple -> output tuple(s)
+	const auto base_columns = chunk.ColumnCount() - projected_input.size();
+
+	DataChunk intermediate_chunk;
```

Comment:
Can we keep all of this state in the `TableInOutLocalState`, instead of allocating/initializing for every chunk we process?